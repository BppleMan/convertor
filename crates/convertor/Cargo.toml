[package]
name = "convertor"
description.workspace = true
authors.workspace = true
version.workspace = true
edition.workspace = true
license.workspace = true
repository.workspace = true

[dependencies]
# 序列化 / 解析 / 配置
toml = { workspace = true, features = ["parse", "display", "serde"] }
serde = { workspace = true, features = ["derive"] }
serde_json = { workspace = true }
serde_yaml = { workspace = true }

# HTTP 客户端 / 网络
reqwest = { workspace = true, features = ["http2", "json", "stream", "rustls-tls"] }
headers = { workspace = true }
httpmock = { workspace = true, features = ["colored"], optional = true }

# URL / 编码 / 工具
percent-encoding = { workspace = true }
url = { workspace = true, features = ["serde"] }
regex = { workspace = true, features = ["unicode-perl"] }
uuid = "1.18.1"

# 异步运行时 / 并发
futures-util = { workspace = true }

# 日志 / 追踪 / 遥测（可选 feature）
tracing = { workspace = true, features = ["attributes"] }
tracing-subscriber = { workspace = true, features = ["default", "env-filter", "json", "local-time"] }
tracing-loki = { workspace = true, default-features = false, features = ["rustls", "compat-0-2-1"] }
tracing-opentelemetry = { workspace = true }

# 指标
opentelemetry = { workspace = true, features = ["trace"] }
opentelemetry_sdk = { workspace = true, features = ["trace", "rt-tokio"] }
opentelemetry-otlp = { workspace = true, features = ["trace", "grpc-tonic"] }

# 加密 / 哈希 / 编码
base64 = { workspace = true }
chacha20poly1305 = { workspace = true, features = ["getrandom", "alloc"] }
rand_core = { workspace = true, features = ["std", "os_rng"] }
rand_chacha = { workspace = true, features = ["default"] }

# 缓存 / 存储 / 数据库
moka = { workspace = true, features = ["future", "logging"] }
redis = { workspace = true, features = ["tokio-rustls-comp", "streams", "connection-manager"] }

# CLI / 工具 / 键入
clap = { workspace = true, features = ["std", "derive"] }

# 错误 / 诊断
color-eyre = { workspace = true, optional = true }
thiserror = { workspace = true }

[dev-dependencies]
# 测试工具
insta = { workspace = true, features = ["filters"] }
serde_json = { workspace = true }

[features]
default = []
testkit = ["httpmock", "color-eyre"]
#otel = [
#    "opentelemetry",
#    "opentelemetry_sdk",
#    "opentelemetry-otlp",
#    "tracing-opentelemetry",
#]

[[test]]
name = "encrypt_test"
path = "tests/encrypt_test.rs"
required-features = ["testkit"]

[[test]]
name = "url_builder_test"
path = "tests/url_builder_test.rs"
required-features = ["testkit"]
