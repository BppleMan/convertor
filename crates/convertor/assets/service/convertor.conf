server {
    listen 443 ssl;
    http2 on;
    server_name convertor.bppleman.cn;

    # SSL 证书
    ssl_certificate     /etc/ssl/certs/bppleman.cn.pem;
    ssl_certificate_key /etc/ssl/private/bppleman.cn.key;

    # SSL 加密推荐配置
    ssl_protocols TLSv1.2 TLSv1.3;
    ssl_prefer_server_ciphers on;
    ssl_ciphers HIGH:!aNULL:!MD5;

    # 专属日志路径
    access_log /var/log/nginx/convertor_access.log main;
    error_log  /var/log/nginx/convertor_error.log warn;

    # 反代设置
    location / {
        proxy_pass http://127.0.0.1:8080;

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_buffering off;

        proxy_connect_timeout 10s;
        proxy_send_timeout    30s;
        proxy_read_timeout    30s;
    }

    # gzip 压缩
    gzip on;
    gzip_types text/plain application/json application/javascript text/css;

    # 安全 header（可按需删改）
    add_header X-Content-Type-Options nosniff;
    add_header X-Frame-Options DENY;
    add_header X-XSS-Protection "1; mode=block";
}
