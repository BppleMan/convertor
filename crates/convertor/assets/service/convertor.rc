#!/sbin/openrc-run

name="convertor"
description="Convertor Service"

# 服务二进制
command="/root/convertor"
command_args="127.0.0.1:8080"
command_user="root"
directory="/root/.convertor"

# 运行模式
command_background=yes
start_stop_daemon_args="--make-pidfile --background"
pidfile="/run/convertor/convertor.pid"

# 可选：日志路径（如需要静音可改为 /dev/null）
output_log="/var/log/${RC_SVCNAME}.out.log"
error_log="/var/log/${RC_SVCNAME}.err.log"

# 环境变量注入方式一：从 conf.d 读取
# 推荐：在 /etc/conf.d/convertor 文件中定义，例如：
#   REDIS_ENDPOINT=redis.example.com
#   APP_ENV=production
# OpenRC 会自动加载
env_file="/etc/conf.d/${RC_SVCNAME}"

depend() {
    need net
}

rc_ulimit="-n 1024"
