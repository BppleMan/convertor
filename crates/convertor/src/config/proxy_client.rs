use clap::ValueEnum;
use serde::{Deserialize, Serialize};
use std::fmt::{Display, Formatter};
use std::str::FromStr;

#[derive(Default, Debug, Copy, Clone, Eq, PartialEq, Hash)]
#[derive(ValueEnum, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ProxyClient {
    #[default]
    Surge,
    Clash,
}

impl ProxyClient {
    pub fn as_str(&self) -> &'static str {
        match self {
            ProxyClient::Surge => "surge",
            ProxyClient::Clash => "clash",
        }
    }

    pub fn variants() -> &'static [Self] {
        &[Self::Surge, Self::Clash]
    }
}

impl Display for ProxyClient {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.as_str())
    }
}

impl FromStr for ProxyClient {
    type Err = String;

    fn from_str(s: &str) -> core::result::Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "surge" => Ok(ProxyClient::Surge),
            "clash" => Ok(ProxyClient::Clash),
            _ => Err(format!("Invalid proxy client: {}", s)),
        }
    }
}
