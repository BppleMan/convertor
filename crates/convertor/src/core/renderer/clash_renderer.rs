use crate::config::client_config::ProxyClient;
use crate::core::profile::clash_profile::ClashProfile;
use crate::core::profile::policy::Policy;
use crate::core::profile::proxy::Proxy;
use crate::core::profile::proxy_group::ProxyGroup;
use crate::core::profile::rule::{ProviderRule, Rule};
use crate::core::profile::rule_provider::RuleProvider;
use crate::core::renderer::{INDENT, Renderer};
use crate::error::RenderError;
use std::fmt::Write;
use tracing::instrument;

type Result<T> = core::result::Result<T, RenderError>;

pub struct ClashRenderer;

impl Renderer for ClashRenderer {
    type PROFILE = ClashProfile;

    fn client() -> ProxyClient {
        ProxyClient::Clash
    }

    #[instrument(skip_all)]
    fn render_profile(profile: &Self::PROFILE) -> Result<String> {
        let mut output = String::new();
        writeln!(output, "{}", Self::render_general(profile)?)?;

        let proxies = Self::render_proxies(&profile.proxies)?;
        writeln!(output, "proxies:")?;
        writeln!(output, "{proxies}")?;

        let proxy_groups = Self::render_proxy_groups(&profile.proxy_groups)?;
        writeln!(output, "proxy-groups:")?;
        writeln!(output, "{proxy_groups}")?;

        let rule_providers = Self::render_rule_providers(&profile.rule_providers)?;
        writeln!(output, "rule-providers:")?;
        writeln!(output, "{rule_providers}")?;

        let rules = Self::render_rules(&profile.rules)?;
        writeln!(output, "rules:")?;
        writeln!(output, "{rules}")?;

        Ok(output)
    }

    #[instrument(skip_all)]
    fn render_general(profile: &Self::PROFILE) -> Result<String> {
        let mut output = String::new();
        writeln!(output, "port: {}", profile.port)?;
        writeln!(output, "socks-port: {}", profile.socks_port)?;
        writeln!(output, "redir-port: {}", profile.redir_port)?;
        writeln!(output, "allow-lan: {}", profile.allow_lan)?;
        writeln!(output, "mode: {}", profile.mode)?;
        writeln!(output, "log-level: {}", profile.log_level)?;
        writeln!(output, r#"external-controller: {}"#, profile.external_controller)?;
        writeln!(output, r#"external-ui: {}"#, profile.external_ui)?;
        if let Some(secret) = &profile.secret {
            writeln!(output, r#"secret: "{secret}""#)?;
        }
        Ok(output)
    }

    fn render_proxy(proxy: &Proxy) -> Result<String> {
        let mut output = String::new();
        write!(output, "{{ ")?;
        write!(output, r#"name: "{}""#, &proxy.name)?;
        write!(output, r#", type: "{}""#, &proxy.r#type)?;
        write!(output, r#", server: "{}""#, &proxy.server)?;
        write!(output, r#", port: {}"#, &proxy.port)?;
        write!(output, r#", password: "{}""#, &proxy.password)?;
        if let Some(udp) = &proxy.udp {
            write!(output, r#", udp: {udp}"#)?;
        }
        if let Some(tfo) = &proxy.tfo {
            write!(output, r#", tfo: {tfo}"#)?;
        }
        if let Some(cipher) = &proxy.cipher {
            write!(output, r#", cipher: {cipher}"#)?;
        }
        if let Some(sni) = &proxy.sni {
            write!(output, r#", sni: "{sni}""#)?;
        }
        if let Some(skip_cert_verify) = &proxy.skip_cert_verify {
            write!(output, r#", skip-cert-verify: {skip_cert_verify}"#)?;
        }
        write!(output, " }}")?;
        Ok(output)
    }

    fn render_proxy_group(proxy_group: &ProxyGroup) -> Result<String> {
        let mut output = String::new();
        write!(output, "{{ ")?;
        write!(output, r#"name: "{}""#, proxy_group.name)?;
        write!(output, r#", type: "{}""#, proxy_group.r#type.as_str())?;
        write!(output, r#", proxies: [ {} ]"#, proxy_group.proxies.join(", "))?;
        write!(output, " }}")?;
        Ok(output)
    }

    fn render_rule(rule: &Rule) -> Result<String> {
        let mut output = String::new();
        write!(output, "{}", rule.rule_type.as_str())?;
        if let Some(value) = &rule.value {
            write!(output, ",{value}")?;
        }
        write!(output, ",{}", Self::render_policy(&rule.policy)?)?;
        Ok(output)
    }

    fn render_rule_for_provider(rule: &Rule) -> Result<String> {
        Self::render_rule(rule)
    }

    fn render_provider_rule(rule: &ProviderRule) -> Result<String> {
        Ok(format!("{},{}", rule.rule_type.as_str(), rule.value))
    }

    #[instrument(skip_all)]
    fn render_rule_providers(rule_providers: &[(String, RuleProvider)]) -> Result<String> {
        let output = rule_providers
            .iter()
            .map(Self::render_rule_provider)
            .map(|line| line.map(|line| format!("{:indent$}{}", "", line, indent = INDENT)))
            .collect::<Result<Vec<_>>>()?
            .join("\n");
        Ok(output)
    }

    fn render_rule_provider(rule_provider: &(String, RuleProvider)) -> Result<String> {
        let (name, rule_provider) = rule_provider;
        Ok(format!(
            r#"{}: {{ type: "{}", url: "{}", path: "{}", interval: {}, size-limit: {}, format: "{}", behavior: "{}" }}"#,
            name,
            rule_provider.r#type,
            rule_provider.url,
            rule_provider.path,
            rule_provider.interval,
            rule_provider.size_limit,
            rule_provider.format,
            rule_provider.behavior
        ))
    }

    fn render_provider_name_for_policy(policy: &Policy) -> String {
        let mut output = if policy.is_subscription {
            "Subscription".to_string()
        } else {
            policy.name.clone()
        };
        match &policy.option {
            Some(option) => {
                output.push('_');
                output.push_str(option.replace('-', "_").as_str());
            }
            None => {
                output.push_str("_policy");
            }
        }
        output
    }
}
