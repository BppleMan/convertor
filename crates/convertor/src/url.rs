pub mod convertor_url;
pub mod query;
pub mod url_builder;
pub mod url_error;
