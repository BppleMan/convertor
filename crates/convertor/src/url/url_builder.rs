use crate::common::encrypt::encrypt;
use crate::config::client_config::ProxyClient;
use crate::config::provider_config::Provider;
use crate::core::profile::policy::Policy;
use crate::core::profile::surge_header::SurgeHeader;
use crate::error::UrlBuilderError;
use crate::url::convertor_url::{ConvertorUrl, ConvertorUrlType};
use crate::url::query::ConvertorQuery;
use serde::{Deserialize, Serialize};
use url::Url;

#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
pub struct UrlBuilder {
    pub secret: String,
    pub enc_secret: String,
    pub client: ProxyClient,
    pub provider: Provider,
    pub server: Url,
    pub sub_url: Url,
    pub enc_sub_url: String,
    pub interval: u64,
    pub strict: bool,
}

impl UrlBuilder {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        secret: impl AsRef<str>,
        enc_secret: Option<String>,
        client: ProxyClient,
        provider: Provider,
        server: Url,
        sub_url: Url,
        enc_sub_url: Option<String>,
        interval: u64,
        strict: bool,
    ) -> Result<Self, UrlBuilderError> {
        let secret = secret.as_ref().to_string();
        let enc_secret = enc_secret
            .map(Ok)
            .unwrap_or_else(|| encrypt(secret.as_bytes(), secret.as_str()))?;
        let enc_sub_url = enc_sub_url
            .map(Ok)
            .unwrap_or_else(|| encrypt(secret.as_bytes(), sub_url.as_str()))?;

        let builder = Self {
            secret,
            enc_secret,
            client,
            provider,
            server,
            sub_url,
            enc_sub_url,
            interval,
            strict,
        };
        Ok(builder)
    }

    pub fn from_convertor_query(
        query: ConvertorQuery,
        secret: impl AsRef<str>,
        client: ProxyClient,
        provider: Provider,
    ) -> Result<Self, UrlBuilderError> {
        let ConvertorQuery {
            server,
            sub_url,
            enc_sub_url,
            interval,
            strict,
            secret: secret_opt,
            enc_secret,
            policy: _,
        } = query;
        let secret = secret_opt.unwrap_or(secret.as_ref().to_string());
        let strict = strict.unwrap_or(true);
        Self::new(
            secret,
            enc_secret,
            client,
            provider,
            server,
            sub_url,
            Some(enc_sub_url),
            interval,
            strict,
        )
    }

    pub fn build_raw_url(&self) -> ConvertorUrl {
        let mut url = self.sub_url.clone();
        url.query_pairs_mut().append_pair("flag", self.client.as_ref());
        ConvertorUrl::raw(url)
    }

    pub fn build_raw_profile_url(&self) -> Result<ConvertorUrl, UrlBuilderError> {
        let query = self.as_profile_query().encode_to_profile_query()?;
        let url = ConvertorUrl::raw_profile(
            self.server.clone(),
            format!(
                "/raw-profile/{}/{}",
                self.client.as_ref().to_ascii_lowercase(),
                self.provider.as_ref().to_ascii_lowercase()
            ),
            query,
        );
        Ok(url)
    }

    pub fn build_profile_url(&self) -> Result<ConvertorUrl, UrlBuilderError> {
        let query = self.as_profile_query().encode_to_profile_query()?;
        let url = ConvertorUrl::profile(
            self.server.clone(),
            format!(
                "/profile/{}/{}",
                self.client.as_ref().to_ascii_lowercase(),
                self.provider.as_ref().to_ascii_lowercase()
            ),
            query,
        );
        Ok(url)
    }

    pub fn build_rule_provider_url(&self, policy: &Policy) -> Result<ConvertorUrl, UrlBuilderError> {
        let query = self.as_rule_provider_query(policy).encode_to_rule_provider_query()?;
        let url = ConvertorUrl::rule_provider(
            policy.clone(),
            self.server.clone(),
            format!(
                "/rule-provider/{}/{}",
                self.client.as_ref().to_ascii_lowercase(),
                self.provider.as_ref().to_ascii_lowercase()
            ),
            query,
        );
        Ok(url)
    }

    pub fn build_sub_logs_url(&self) -> Result<ConvertorUrl, UrlBuilderError> {
        let query = self.as_sub_logs_query().encode_to_sub_logs_query()?;
        let url = ConvertorUrl::sub_logs(self.server.clone(), "/api/sub-logs", query);
        Ok(url)
    }

    // 构造专属 Surge 的订阅头
    pub fn build_surge_header(&self, r#type: ConvertorUrlType) -> Result<SurgeHeader, UrlBuilderError> {
        let url = match r#type {
            ConvertorUrlType::Raw => self.build_raw_url(),
            ConvertorUrlType::RawProfile => self.build_raw_profile_url()?,
            ConvertorUrlType::Profile => self.build_profile_url()?,
            _ => return Err(UrlBuilderError::UnsupportedUrlType(r#type)),
        };
        Ok(SurgeHeader::new(url, self.interval, self.strict))
    }
}

impl UrlBuilder {
    pub fn as_profile_query(&self) -> ConvertorQuery {
        ConvertorQuery {
            server: self.server.clone(),
            sub_url: self.sub_url.clone(),
            enc_sub_url: self.enc_sub_url.clone(),
            interval: self.interval,
            strict: Some(self.strict),
            policy: None,
            secret: None,
            enc_secret: None,
        }
    }

    pub fn as_rule_provider_query(&self, policy: &Policy) -> ConvertorQuery {
        let mut query = self.as_profile_query();
        query.policy = Some(policy.clone());
        query
    }

    pub fn as_sub_logs_query(&self) -> ConvertorQuery {
        let mut query = self.as_profile_query();
        query.secret = Some(self.secret.clone());
        query.enc_secret = Some(self.enc_secret.clone());
        query
    }
}

pub trait HostPort {
    fn host_port(&self) -> Option<String>;
}

impl HostPort for Url {
    fn host_port(&self) -> Option<String> {
        match (self.host_str(), self.port()) {
            (Some(host), Some(port)) => Some(format!("{host}:{port}")),
            (Some(host), None) => Some(host.to_string()),
            _ => None,
        }
    }
}
