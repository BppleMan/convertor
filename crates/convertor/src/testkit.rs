use crate::common::config::ConvertorConfig;
use crate::common::config::provider_config::ProviderConfig;
use crate::common::config::proxy_client_config::ProxyClient;
use crate::common::encrypt::nonce_rng_use_seed;
use crate::common::once::{init_backtrace, init_log};
use crate::core::profile::policy::Policy;
use crate::url::url_builder::HostPort;
use color_eyre::Report;
use httpmock::Method::{GET, POST};
use httpmock::MockServer;
use std::path::{Path, PathBuf};
use strum::VariantArray;
use url::Url;

pub fn init_test() -> PathBuf {
    let base_dir = Path::new(env!("CARGO_MANIFEST_DIR")).join("test-assets");
    init_backtrace();
    init_log(None);
    nonce_rng_use_seed([0u8; 32]);
    base_dir
}

pub async fn start_mock_provider_server(config: &mut ConvertorConfig) -> Result<(), Report> {
    for (_, config) in config.providers.iter_mut() {
        config.start_mock_provider_server().await?;
    }
    Ok(())
}

pub(crate) trait MockServerExt {
    async fn start_mock_provider_server(&mut self) -> Result<MockServer, Report>;
}

impl MockServerExt for ProviderConfig {
    async fn start_mock_provider_server(&mut self) -> Result<MockServer, Report> {
        let mock_server = MockServer::start_async().await;

        // 将订阅地址导航至 mock server 的 /subscription 路径
        let subscribe_url_path = "/subscription";
        let token = "bppleman";

        self.sub_url =
            Url::parse(&mock_server.url(format!("{subscribe_url_path}?token={token}"))).expect("不合法的订阅地址");
        self.api_config.host = Url::parse(&mock_server.url("/")).expect("不合法的 API 地址");

        mock_server
            .mock_async(|when, then| {
                when.method(POST).path(self.api_config.login_path());
                let body = serde_json::json!({
                    "data": {
                        "auth_data": "mock_auth_token"
                    }
                });
                then.status(200)
                    .body(serde_json::to_string(&body).unwrap())
                    .header("Content-Type", "application/json");
            })
            .await;

        mock_server
            .mock_async(|when, then| {
                when.method(GET).path(self.api_config.get_sub_path());
                let body = serde_json::json!({
                    "data": {
                        "subscribe_url": mock_server.url(format!("{subscribe_url_path}?token={token}")),
                    }
                });
                then.status(200)
                    .body(serde_json::to_string(&body).unwrap())
                    .header("Content-Type", "application/json");
            })
            .await;

        mock_server
            .mock_async(|when, then| {
                when.method(POST).path(self.api_config.reset_sub_path());
                let body = serde_json::json!({
                    "data": mock_server.url(format!("{subscribe_url_path}?token=reset_{token}")),
                });
                then.status(200)
                    .body(serde_json::to_string(&body).unwrap())
                    .header("Content-Type", "application/json");
            })
            .await;

        // hook mock server 的 /subscription 路径，返回相应的 mock 数据
        let sub_host = self.sub_url.host_port()?;
        for client in ProxyClient::VARIANTS {
            mock_server
                .mock_async(|when, then| {
                    when.method(GET)
                        .path(subscribe_url_path)
                        .query_param("flag", client.as_ref())
                        .query_param("token", token);
                    let body = mock_profile(*client, &sub_host);
                    then.status(200)
                        .body(body)
                        .header("Content-Type", "text/plain; charset=utf-8");
                })
                .await;
        }

        Ok(mock_server)
    }
}

pub fn mock_profile(client: ProxyClient, sub_host: impl AsRef<str>) -> String {
    match client {
        ProxyClient::Surge => include_str!(concat!(
            env!("CARGO_MANIFEST_DIR"),
            "/test-assets/surge/mock_profile.conf"
        )),
        ProxyClient::Clash => include_str!(concat!(
            env!("CARGO_MANIFEST_DIR"),
            "/test-assets/clash/mock_profile.yaml"
        )),
    }
    .replace("{sub_host}", sub_host.as_ref())
}

pub fn policies() -> [Policy; 7] {
    [
        Policy::subscription_policy(),
        Policy::new("BosLife", None, false),
        Policy::new("BosLife", Some("no-resolve"), false),
        Policy::new("BosLife", Some("force-remote-dns"), false),
        Policy::direct_policy(None),
        Policy::direct_policy(Some("no-resolve")),
        Policy::direct_policy(Some("force-remote-dns")),
    ]
}
