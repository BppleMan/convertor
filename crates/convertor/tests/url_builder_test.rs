use convertor::config::proxy_client::ProxyClient;
use convertor::core::renderer::Renderer;
use convertor::core::renderer::clash_renderer::ClashRenderer;
use convertor::init_test;
use convertor::testkit::policies;
use convertor::url::url_builder::UrlBuilder;
use url::Url;

fn url_builder(client: ProxyClient) -> color_eyre::Result<UrlBuilder> {
    let server = Url::parse("http://127.0.0.1:8080")?;
    let sub_url = Url::parse("https://localhost/subscription?token=bppleman")?;
    let secret = "bppleman_secret";
    let url_builder = UrlBuilder::new(secret, None, client, server.clone(), sub_url.clone(), None, 86400, true)?;
    Ok(url_builder)
}

#[test]
fn test_url_builder_surge() -> color_eyre::Result<()> {
    init_test!();
    let url_builder = url_builder(ProxyClient::Surge)?;
    let raw_url = url_builder.build_raw_url();
    insta::assert_snapshot!(raw_url.to_string(), @"https://localhost/subscription?token=bppleman&flag=surge");

    let raw_profile_url = url_builder.build_raw_profile_url()?;
    insta::assert_snapshot!(raw_profile_url.to_string(), @"http://127.0.0.1:8080/raw-profile/surge?interval=86400&strict=true&sub_url=qDbvzIt3DcfaQVl8UVdIjXck4D-42Eo3c6g0ZR2cc7lpxUFnkUEWW0fTRfMAmY3yU3f-ESJYD93o5YDKtEzSe1ATkzfrq9RxPdh7fMif0IOZXScDcg");

    let profile_url = url_builder.build_profile_url()?;
    insta::assert_snapshot!(profile_url.to_string(), @"http://127.0.0.1:8080/profile/surge?interval=86400&strict=true&sub_url=qDbvzIt3DcfaQVl8UVdIjXck4D-42Eo3c6g0ZR2cc7lpxUFnkUEWW0fTRfMAmY3yU3f-ESJYD93o5YDKtEzSe1ATkzfrq9RxPdh7fMif0IOZXScDcg");

    let policies = policies();
    for policy in policies {
        let ctx = format!(
            "test_url_builder_surge_{}",
            ClashRenderer::render_provider_name_for_policy(&policy)
        );
        let rule_provider_url = url_builder.build_rule_provider_url(&policy)?;
        insta::assert_snapshot!(ctx, rule_provider_url.to_string());
    }
    Ok(())
}

#[test]
fn test_url_builder_clash_boslife() -> color_eyre::Result<()> {
    init_test!();
    let url_builder = url_builder(ProxyClient::Clash)?;
    let raw_url = url_builder.build_raw_url();
    insta::assert_snapshot!(raw_url.to_string(), @"https://localhost/subscription?token=bppleman&flag=clash");

    let raw_profile_url = url_builder.build_raw_profile_url()?;
    insta::assert_snapshot!(raw_profile_url.to_string(), @"http://127.0.0.1:8080/raw-profile/clash?interval=86400&strict=true&sub_url=qDbvzIt3DcfaQVl8UVdIjXck4D-42Eo3c6g0ZR2cc7lpxUFnkUEWW0fTRfMAmY3yU3f-ESJYD93o5YDKtEzSe1ATkzfrq9RxPdh7fMif0IOZXScDcg");

    let profile_url = url_builder.build_profile_url()?;
    insta::assert_snapshot!(profile_url.to_string(), @"http://127.0.0.1:8080/profile/clash?interval=86400&strict=true&sub_url=qDbvzIt3DcfaQVl8UVdIjXck4D-42Eo3c6g0ZR2cc7lpxUFnkUEWW0fTRfMAmY3yU3f-ESJYD93o5YDKtEzSe1ATkzfrq9RxPdh7fMif0IOZXScDcg");

    let policies = policies();
    for policy in policies {
        let ctx = format!(
            "test_url_builder_clash_{}",
            ClashRenderer::render_provider_name_for_policy(&policy)
        );
        let rule_provider_url = url_builder.build_rule_provider_url(&policy)?;
        insta::assert_snapshot!(ctx, rule_provider_url.to_string());
    }
    Ok(())
}
