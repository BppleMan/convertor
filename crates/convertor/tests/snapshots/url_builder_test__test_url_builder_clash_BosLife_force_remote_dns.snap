---
source: crates/convertor/tests/url_builder_test.rs
expression: rule_provider_url.to_string()
---
http://127.0.0.1:8080/rule-provider/clash?interval=86400&policy[name]=BosLife&policy[option]=force-remote-dns&policy[is_subscription]=false&sub_url=qDbvzIt3DcfaQVl8UVdIjXck4D-42Eo3c6g0ZR2cc7lpxUFnkUEWW0fTRfMAmY3yU3f-ESJYD93o5YDKtEzSe1ATkzfrq9RxPdh7fMif0IOZXScDcg
